V4 update test framework for distributions random method 2nd attempt

matteo-pallini opened the pull request:
The recent distributions refactoring (#4508 and #4548) moves the logic to 
get a random variable from a distribution to aesara. 
This relies on numpy and scipy random variables implementation.
Given this change the main thing worth testing on the PyMC side is that 
the PyMC parametrization is sensible given the one on the Aesara side
(effectively the numpy/scipy one)

More details can be found on issue #4554

- this is the second attempt for doing this change. The first one can be found here: 
https://github.com/pymc-devs/pymc3/pull/4580

+ [x] what are the (breaking) changes that this PR makes? No breaking changes, only fixing currently broken tests
+ [x] are the changes—especially new features—covered by tests and docstrings? This change is only affecting tests
+ [x] Is there any other test not relevant anymore given the refactoring?

matteo-pallini committed:
Update tests following distributions refactoring

The distributions refactoring moves the random variable sampling to
aesara. This relies on numpy and scipy random variables implementation.
So, now the only thing we care about testing is that the parametrization
on the PyMC side is sendible given the one on the Aesara side
(effectively the numpy/scipy one)

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

matteo-pallini committed:
Change tests for more refactored distributions.

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

matteo-pallini committed:
Change tests for refactored distributions

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

matteo-pallini committed:
Remove tests for random variable samples shape and size

Most of the random variable logic has been moved to aesara, as well as
most of the relative tests. More details can be found on issue #4554

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -497,43 +424,136 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+        expected_params = [("n", expected_n), ("p", expected_p)]
+        self._pymc_params_match_rv_ones(params, expected_params, pm.NegativeBinomial)
+
+    def test_bernoulli(self):
```
We should also test for logit_p parametrization, even if it was not being tested before.

matteo-pallini replied:
`logit_p` is currently not being used by the method. I added a test which expects a raise. Let me know whether you had something else in mind

ricardoV94 replied:
Ah, I thought it was already refactored. Then maybe let's put it as an xfail which should pass once it is refactored.

matteo-pallini replied:
I think that as long as the implementation and the expected behaviors are not known it may make sense to keep an expected raise rather than an xfail on something undefined. If I were to define the test behavior, I should probably also define the method behavior

ricardoV94 replied:
Sure, let's keep it like this then :)

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -497,43 +424,136 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+        params = [("mu", 4)]
+        self._pymc_params_match_rv_ones(params, params, pm.Poisson)
+
+    def test_mv_distribution(self):
```
Should be called mvnormal or mv_normal, as there will be several mvs in the future.

matteo-pallini replied:
Good point, fixed

ricardoV94 reviewed (commented):
Looks like a good start. We should also check that `.eval()` actually works even if just for a single sample.

I think @brandonwillard had some code in the original issue asserting that the numbers generated by our distribution and the scipy one were the same when given the same random seeds. That feels like a stronger test. 

PS: Checking that a dozen or so samples match should be enough as we already asserted the right parameters are being used.

What do you think?

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -497,43 +424,136 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+        params = [("alpha", 2.0), ("beta", 5.0)]
+        self._pymc_params_match_rv_ones(params, params, pm.Cauchy)
+
+    def test_half_cauchy(self):
```
This one is failing because we have not implemented alpha parameter on the pymc3 side. So it must always be zero, as you can see in this line: https://github.com/pymc-devs/pymc3/blob/23cffdcc6a5ed161ece0156b17a4bb9648fec3a7/pymc3/distributions/continuous.py#L2266

matteo-pallini replied:
Oh, I missed that! thanks for pointing it out

matteo-pallini committed:
Fix test for half cauchy, renmae mv normal tests and add test for
Bernoulli

ricardoV94 commented:
The exponential and gamma should now pass after https://github.com/pymc-devs/pymc3/pull/4576

matteo-pallini commented:
> The exponential and gamma should now pass after #4576

Lovely, thanks for flagging

matteo-pallini committed:
Add test checking PyMC samples match the aesara ones

Also mark test_categorical as expected to fail due to bug on aesara
side. The bug is going to be fixed with 2.0.5 release, so we need to
bump the version for categorical and the test to pass.

matteo-pallini committed:
Move Aesara to 2.0.5 to include Gumbel distribution

matteo-pallini committed:
Enamble exponential and gamma tests following bug-fix

matteo-pallini commented:
> Looks like a good start. We should also check that `.eval()` actually works even if just for a single sample.
> 
> I think @brandonwillard had some code in the original issue asserting that the numbers generated by our distribution and the scipy one were the same when given the same random seeds. That feels like a stronger test.
> 
> PS: Checking that a dozen or so samples match should be enough as we already asserted the right parameters are being used.
> 
> What do you think?

Sounds good. It's definitely a more robust test. Let me know if what added is in line with what you had in mind

matteo-pallini committed:
Enable categorical test following aesara version bump to 2.0.5 and relative bug-fix

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -498,43 +425,177 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+
+    def test_beta_mu_sigma(self):
+        params = [("mu", 0.5), ("sigma", 0.25)]
+        expected_alpha, expected_beta = pm.Beta.get_alpha_beta(mu=params[0][1], sigma=params[1][1])
```
Small nitpick, would it be a lot of trouble to write the tests expecting a dictionary of `{"parameter_name": parameter_value}` instead of the list of tuples? I think it makes it easier to write / read (instead of the `params[0][1]` kind of logic), and it seems that in the test function you are converting it to a dict anyway. Dictionaries are ordered from 3.6, which is what we support, so order wouldn't be an issue.

ricardoV94 replied:
Also maybe with this refactoring we can avoid the `params, list(params)` in the function call, which is not very intuitive.

matteo-pallini replied:
Good idea, about using a dict, given its current ordered nature! I live in the past :-). yeah `list(params)` is very ugly, sorry for that, worse case I will replace it with `.copy`

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -498,43 +425,177 @@ class TestMoyal(BaseTestCases.BaseTestCase):
-        self, shape, with_vector_params=False, **kwargs
-    ):  # don't transform categories
-        return super().get_random_variable(shape, with_vector_params=False, **kwargs)
+    def _compare_pymc_sampling_with_aesara_one(
```
Small nitpick, I would make `pymc_dist` the first argument

ricardoV94 replied:
Even smaller nitpick no need for `_one(s)` suffix.

ricardoV94 commented:
I left some small comments. Overall I think it looks fine and should definitely cover the pymc <- -> aesara parametrizations. After this PR we might still need something extra for the random ops that are implemented on our side, but that might be better done separately.

ricardoV94 commented:
Also, we can remove these lines from the previous PR: https://github.com/pymc-devs/pymc3/blob/7a7179276f8f60470dacc496d3c2add4b636e4a2/pymc3/tests/test_distributions_random.py#L1773-L1803

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -498,43 +425,177 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+    def _pymc_sample_matches_aeasara_rv_one(
+        self, pymc_dist_output, pymc_dist, expected_aesara_params, size, decimal
+    ):
+        sample = pymc_dist.rv_op(
```
I am not sure of what I am about to say, but here it goes:

We might want to go a bit more low level here and compare with the samples from the underlying "reference" number generating function (say `rng.distribution` or `scipy.stats.distribution.rvs()`) directly. Otherwise we are just comparing that the call from the `random op` and `pymc_dist.eval()` are the same, which is well covered by the `_pymc_params_match_aesara...`. 

Silly example: if we changed our exponential to return the outcome of `rng.exponential(mu+1)` instead of `rng.exponential(mu)` this test would not capture the accidental change. However, if we pass the reference function explicitly and call it with the "expected parameters" and the same seed, we would find that the outcomes don't match anymore.

As an example of how to define the reference random distribution see here: https://github.com/pymc-devs/pymc3/blob/b8edd5671cecf719d7059e3d728182db66c209bd/pymc3/tests/test_distributions_random.py#L819-L827

Basically duplicating the code that generates the random numbers...

It doesn't provide much more guarantees, but could make this test more useful for the remaining distributions that have not been refactored yet.

Does this make sense at all to you?

matteo-pallini replied:
I agree with you in principle, but I am not sure what's the best way to achieve that outcome in practice. Below my 2 cents.

The way I see it Aesara is responsible for generating the correct distribution given the rv class name. Indirectly Aesara is also responsible for testing that this is the case. PyMC relies on Aesara doing a good job at this, in the same way that Aesara relies on Scipy/Numpy doing a good job at implementation correctly the various distributions' methods (ie I assume these are not tested for correctness in Aesara, maybe I am wrong though).

I agree with you that adding a lower level test here will increase our confidence that everything works as expected, but I feel that probably these tests should be added to Aesara rather than here. Particularly so, given that the PyMC dev team owns that library, so this should be easy to achieve.

Also, if we decide that we want to have this extra level of safety in PyMC we should probably not rely on knowing Aesara implementation. So, rather than seeding Aesara and our random generators and checking for exact matches in a small sample, we should probably still rely on Kolmogorov Smirnov and Chi-Square tests. ie I wouldn't make the PyMC tests rely on Aesara current implementation (`np/scipy`), which may change in the future, making the tests fail, but still lead to the same result (ie right distribution).

To make sure I get your point fully:
> Silly example: if we changed our exponential to return the outcome of rng.exponential(mu+1) instead of rng.exponential(mu) this test would not capture the accidental change

~Wouldn't this be caught by the parameterization test? like any bug on our side that changes the `dist` arguments in unexpected ways?~

> It doesn't provide much more guarantees, but could make this test more useful for the remaining distributions that have not been refactored yet.

I guess that you would like each distribution to have a test similar to that, right?

matteo-pallini replied:
> We might want to go a bit more low level here and compare with the samples from the underlying "reference" number generating function (say rng.distribution or scipy.stats.distribution.rvs()) directly. 

Btw if you think a straightforward approach like the one you mentioned would work fine, I am happy to go for it

ricardoV94 replied:
Thanks for your input. I agree that for the aesara ops it's a complete overkill. I was hinting at the future random ops that will be implemented mostly on the PyMC3 side. If this test framework would also cover well these future cases it could be worthwhile, even if for the few ops that come from aesara it is indeed an overkill.

But maybe let's leave this as is and tackle future changes if they prove necessary later.

matteo-pallini replied:
>I was hinting at the future random ops that will be implemented mostly on the PyMC3 side

Oh, I see, I didn't know that. What's the best way to stay up to date with this kind of knowledge?

Is this something similar to what you had in mind (haven't ran it yet, so there maybe bugs/typos)? or would you still test it through `pm.Exponential.dist(...)`
```python
from aesara.tensor.random import basic as a_rv_ops

class TestRandomVariableDistribution(SeededTest):

    def _compare_pymc_sampling_with_aesara(self, rv_op_dist, rv_op_params, ref_dist, ref_params, decimal=6, size=15):
        pymc_dist_output = rv_op_dist(*rv_op_params.values(), size=size, rng=aesara.shared(self.get_random_state()))
        sample = ref_dist(self.get_random_state(), **ref_params, size=15)
        assert_array_almost_equal(pymc_dist_output.eval(), sample, decimal=decimal)

    def test_exponential_distribution(self):
        rv_op_params, ref_params = {"lam": 10.}, {"scale": 10.}
        rv_op_dist, ref_dist = a_rv_ops.exponential, np.random.RandomState.exponential
        self._compare_pymc_sampling_with_aesara(rv_op_params, rv_op_params, ref_dist, ref_params)
```

I guess that for now I can only add these tests for rv ops defined in aesara.

matteo-pallini replied:
Also, at this point I feel like the `_compare_pymc_sampling_with_aesara_one` method resulted from me misinterpreting the following `asserting that the numbers generated by our distribution and the scipy one were the same when given the same random seeds. That feels like a stronger test.`. Would it make sense to remove it?

ricardoV94 replied:
> Is this something similar to what you had in mind (haven't ran it yet, so there maybe bugs/typos)? or would you still test it through pm.Exponential.dist(...)

Yes exactly, does it make sense? I would still test through the pymc.dist as users won't be using random ops directly, unless you think this way is significantly easier or there are advantages that I am ignoring.

ricardoV94 replied:
> I guess that for now I can only add these tests for rv ops defined in aesara.

If you want to consider the future ops, you can look at the gumbel random op that was implemented recently in aesara as that calls scipy.stats directly https://github.com/pymc-devs/aesara/pull/354. Most future random ops in pymc will follow that format (ie 1 to 1 correspondence to a scipy distribution rvs call). 

I think the gumbel was already merged into the v4 branch but without tests so you call actually include it in this PR if that helps.

For something a tiny bit more complicated you can check the discrete weibull that I refactored in the draft PR https://github.com/pymc-devs/pymc3/pull/4615. That makes use of random routines in numpy or scipy but is not as simple as calling a pre defined distribution.

I can merge that draft soon if that helps here. I was waiting to have this PR ready first but either way works.

If your tests could cover those 3 types of random ops (calling numpy directly, calling scipy directly, or building upon these more indirectly) it would probably fit most needs.

ricardoV94 replied:
> What's the best way to stay up to date with this kind of knowledge?

It's not always easy, even for core developers as things are moving quite fast right now.

In general some info can be gathered from the meeting minutes: https://github.com/pymc-devs/pymc3/wiki/Project-Meeting-Minutes

Or the wiki more broadly.

Hopefully things will become more predictable soon! It's always fine to ask of course :)

Sorry for the confusion and again thanks for your help and patience!

matteo-pallini replied:
Cool, thanks for the comments and pointing me to the info about where to find the meeting minutes.

I refactored the tests. Now it should be easy to add any extra tests for each distribution. I also added a test comparing the `.dist` outcome with the one coming from the reference distribution. For now it supports scipy and numpy distributions, it shouldn't be too hard to use a different callable to test a custom distribution (like the discrete Weibull you mentioned).

Let me know what you think about the new format. Also, sorry for asking you to re-review the code for this PR basically from scratch. Hopefully we're almost there

ricardoV94 commented:
I think we are pretty much ready after you complete those final refactorings. Thanks for your work and patience.

Is there anything else you think should be changed?

matteo-pallini commented:
> I think we are pretty much ready after you complete those final refactorings. Thanks for your work and patience.
> 
> Is there anything else you think should be changed?

Happy to cover `If this test framework would also cover well these future cases it could be worthwhile` as well. Let's agree on the right way of doing it, if you are happy to. I left a comment with a potential approach on the thread above. I should probably thank you for your patience :-)

I should have addressed all the refactoring comments you left

matteo-pallini committed:
Few small cosmetic changes:
- replace list of tuples with dict
- rename 1 method
- move pymc_dist as first argument in function call
- replace list(params) with params.copy()

matteo-pallini committed:
Remove redundant tests

matteo-pallini committed:
Further refactoring

The refactoring should make it possible testing both the distribution
parametrization and sampled values according to need, as well as any
other future test. More details on PR #4608

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -492,43 +421,316 @@ class TestMoyal(BaseTestCases.BaseTestCase):
-class TestCategorical(BaseTestCases.BaseTestCase):
-    distribution = pm.Categorical
-    params = {"p": np.ones(BaseTestCases.BaseTestCase.shape)}
+class TestRandomVariableDistribution(SeededTest):
```
1. I realized the test above `BaseTestCases` is using `pm.Model()` and `pm.Distribution(name, *args, **kwargs)` instead of `pm.Distribution.dist()`. That sounds like a better approach as that is **actually** how users will interact with distributions. The `.dist()` syntax is rarely used and may actually be dropped in the future. I am sorry for nudging you towards `.dist()`, do you think the tests could work equally well with the other approach?

https://github.com/pymc-devs/pymc3/blob/e4901eeada712cd1f64189a6acc01fcab84f1fc2/pymc3/tests/test_distributions_random.py#L148

2. Do you think we should merge the other two types of tests currently in this file within this general framework? I am talking about the test mentioned in point 1. above which explicitly deals with shapes, and the two older tests `pymc3_random`/ `pymc3_random_discrete` that perform a statistical comparison between a large number of samples of a pymc3 and a reference distribution, to see if they match. The shape test will be important for all ops implemented on our side, while the later two will only be rarely used for distributions that don't have a scipy / numpy direct counterpart but can still be compared to transformations / specific parametrizations of other random distributions (see https://github.com/pymc-devs/pymc3/pull/4596 for an example)

3. I am not sure whether the syntax with the dictionaries and list of tests is the best, but I see why you went with it.

ricardoV94 replied:
Alternatively these tests can live separately from the other two types of tests, in which case I would rename this class of tests to something like `test_pymc3_randomOp_parametrization`

matteo-pallini replied:
1. ok
2. yes, that's the idea. I can add the one dealing with shapes for the gumbel distribution (so that it can be used as a reference). Regarding `pymc3_random` / `pymc3_random_discrete` tests, I guess that the outcome the test is trying to achieve should be the same of `_check_pymc_dist_matches_expected`, so possibly this should be changed to handle non-exact comparisons? after this change the user would specify in `params` whether it's not an exact comparison and whether it's a discrete or continuous distribution and then the right way to compare the two distributions will be picked accordingly (ie current `_check_pymc_dist_matches_expected` or `pymc3_random` or `pymc3_random_discrete`)? wdyt?
3. `I am not sure whether the syntax with the dictionaries and list of tests is the best` me neither, I don't find it particularly elegant tbh. Still, it allows for having only one test per distribution and leaves the flexibility to specify what checks to run. I couldn't think of any other design that had these two benefits and didn't end up being very verbose, or required to draw samples (or run eval()) multiple times. I think that this approach is probably only acceptable for tests (little need for modularity/reusability) Happy to replace it with a better one, if you already have one in mind

matteo-pallini replied:
3. a possibly more pleasant-to-read approach would be creating a base class similar to `BaseTestCase` that defines all the tests and utils needed. Subclass for each distribution, define in each subclass the appropriate parameters individually as class variables (similar to `BaseTestCase`) and have a list of callables containing the tests to ran (same as `tests_to_run` now) as another class variable. This is probably much nicer actually than the current one

ricardoV94 replied:
>  pymc3_random / pymc3_random_discrete tests, I guess that the outcome the test is trying to achieve should be the same of _check_pymc_dist_matches_expected, so possibly this should be changed to handle non-exact comparisons?

On a second thought, I wouldn't bother integrating those two. We might actually not find much use for them. If they were to be integrated I would make them a separate tests from `_check_pymc_dist_matches_expected`, something like `check_draws_match_statistical_test` or something. In the past we were only using them to wastefully compare with the scipy functions directly or with the same random code in our distributions, both of which are better accomplished by your new exact tests. 

The shape test is definitely useful though!

ricardoV94 replied:
> yes, that's the idea. I can add the one dealing with shapes for the gumbel distribution (so that it can be used as a reference) 

Maybe add it to the `Beta` as we are using a custom `rng_fn`. The `Gumbel` is a direct aesara op so it's not the "best" example. 

https://github.com/pymc-devs/pymc3/blob/dc8f5329e847a463be0195aeee23b4d6bf61f2e7/pymc3/distributions/continuous.py#L1080

ricardoV94 replied:
>  a possibly more pleasant-to-read approach would be creating a base class similar to `BaseTestCase` that defines all the tests and utils needed. Subclass for each distribution, define in each subclass the appropriate parameters individually as class variables (similar to `BaseTestCase`) and have a list of callables containing the tests to ran (same as `tests_to_run` now) as another class variable. This is probably much nicer actually than the current one

Sounds good! 

The other more traditional alternative is to make use of *args and **kwargs, and just pass them to the downstream functions.

brandonwillard replied:
> 1. The `.dist()` syntax is rarely used and may actually be dropped in the future.

Yes, and we should create an issue for that.  It should be a fairly simple change to make.

Also, we need to do something about the mandatory name argument.  We definitely can and should remove the unique name requirement, but that's a separate issue.

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -492,43 +421,316 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+        ):
+            assert_almost_equal(expected_value, actual_variable.eval(), decimal=decimal)
+
+    def _get_inputs_from_apply_node_outputs(self, outputs: TensorVariable) -> List[Any]:
```
This might not justify the separate function, unless it starts to be used in more than one type of test

matteo-pallini replied:
Yeah, I tend to prefer using a lot of small functions rather than a few big ones in which the core logic is buried in a lot of code that is only massaging data. I can remove it

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -492,43 +421,316 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+            "expected_rv_op_params": {"mu": 1.5, "beta": 3.0},
+            "expected_dist": self._get_scipy_distribution("gumbel_r"),
+            "expected_dist_params": {"loc": 1.5, "scale": 3.0},
+            "size": 15,
```
Any reason why this needed size explicitly?

matteo-pallini replied:
I guess I can remove it, given that the default value is 15 anyway. it's not needed, but it shows very explicitly how to define the size

ricardoV94 replied:
I think it's fine to leave it out. Specially since the BaseTests parametrize size separately.

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -492,43 +421,316 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+            params, [self._check_pymc_params_match_rv_op, self._check_pymc_dist_matches_expected]
+        )
+
+    def test_normal(self):
```
We should add tests for `tau` and `sd` parametrizations as well. Same for the HalfNormal

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -492,43 +421,316 @@ class TestMoyal(BaseTestCases.BaseTestCase):
-    def test_dirichlet_random_shape(self, shape, size):
-        out_shape = to_tuple(size) + to_tuple(shape)
-        assert pm.Dirichlet.dist(a=np.ones(shape)).random(size=size).shape == out_shape
+    def _check_pymc_dist_matches_expected(
```
Maybe rename to `_check_pymc_draws_match_expected`

ricardoV94 reviewed (commented):
Thanks for the updates. I left some general comments, let me know what you think :)

ricardoV94 commented:
Sounds like you are on the right track.

I would focus on integrating the `BaseTest` with your framework and give it a more explicit name to make it obvious it is testing shapes of draws. If you find the inheriting class logic nicer, that would certainly be a plus.

After this it's perhaps time to get a second review, merge, and start using it with the newly refactored distributions. It will be more obvious whether more types of tests are needed or if these suffice.

matteo-pallini commented:
> I would focus on integrating the BaseTest with your framework and give it a more explicit name to make it obvious it is testing shapes of draws.

I noticed that currently v4 treats the `shape` argument as deprecated, while #4625 is going to reintroduce it. Would it make sense to merge this PR as it is and then incorporate BaseTest to the new framework once #4625 has been merged? alternatively I guess it may make sense to rebase on that before incorporating BaseTest. @ricardoV94 what do you think?

brandonwillard commented:
> I noticed that currently v4 treats the `shape` argument as deprecated, while #4625 is going to reintroduce it.

No code in `v4` should use the `shape` argument; it's only there for backward compatibility.

ricardoV94 commented:
The current BaseTest (prior to this PR) was already working with size. We should keep testing size only. The tests about whether shape is converting nicely to size will be separated and don't need to be done for every distribution. If you can refactor the current size tests then I think this PR will be ready to go.

brandonwillard commented:
> The tests about whether shape is converting nicely to size will be separated and don't need to be done for every distribution.

Exactly.

matteo-pallini committed:
Add size tests to new rv testing framework

matteo-pallini committed:
Add tests for multivariate and for univariate multi-parameters

matteo-pallini committed:
remove test already covered in aesara

matteo-pallini committed:
fix few names

matteo-pallini commented:
I integrated BaseTest into the new framework + some minor renaming and refactoring. I think the PR should be ready for the pre-merge review. Let me know what you think

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -491,43 +414,364 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+    tests_to_run = ["check_pymc_params_match_rv_op"]
+
+
+class TestBetaAlphaBetaDistribution(BaseTestDistribution):
```
Would add a size and matches draws test here as we are using our slightly customized method beta_clipped something

ricardoV94 replied:
Would also be a good example of how to test custom random generators that are not just direct calls to numpy/scipy functions

matteo-pallini replied:
yeah, the Discrete Weibull test should be a good example of handling custom random generators. I haven't had time to finish it yesterday. I should be able to push it today, together with the other test for the `TestBetaAlphaBetaDistribution`

ricardoV94 replied:
No rush, and great work ;)

matteo-pallini replied:
Good comments, I should have addressed them all

ricardoV94 reviewed (commented):
Looks pretty neat. Left a suggestion for the beta test. 

Do you mind rebasing and adding the test for the Discrete Weibull that was recently refactored?

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -491,43 +414,364 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+        for test_name in self.tests_to_run:
+            self.run_test(test_name)
+
+    def run_test(self, test_name):
```
This can just be a class dictionary with name:class methods, no?

matteo-pallini replied:
I think that it may result uglier than how it is at the moment. 

I would need to convert all the `_check` methods in class methods and have the class dictionary being defined after these, so that it could refer to them. But also the class dictionary will need to be defined right before `test_distribution`, so that this could also refer to it. As per below
```python
class BaseTestDistribution:

    @classmethod
    def _check_1(cls):
        assert True
    
    avaliable_tests = {
        "check_1": _check_1
    }
    
    def test_distributions(self):
        self.avaliable_tests["check_1"]()

```
If we don't want to re-define the dictionary for each single test-class, or if we want to get rid of unnecessary code and make it a bit more implicit we could simply do a bit of renaming and something along these lines:
```python
def run_test(self, test_name):
    getattr(self, test_name)()
````

ricardoV94 replied:
Thanks for the explanation. I agree with your point. I think the last approach is nicer than the current one, and you can drop the `run_test` function all together in that case as well, no?

```python
for test_name in self.tests_to_run:
  getattr(self, test_name)()
```

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -491,43 +414,364 @@ class TestMoyal(BaseTestCases.BaseTestCase):
-        assert pm.Categorical.dist(p=p).random().shape == (3, 7)
-        assert pm.Categorical.dist(p=p).random(size=4).shape == (4, 3, 7)
+    def _check_pymc_params_match_rv_op(self) -> None:
+        try:
```
Do we need this `try / except` block? The test will just fail if this happens, and I don't see why it would happen in the first place (?)

matteo-pallini replied:
true, removing it

ricardoV94 commented:
Also there seems to still be an old Beta test above: https://github.com/drabbit17/pymc3/blob/fe3b206fac51cdfdef35466a37337c0646ba0bc0/pymc3/tests/test_distributions_random.py#L287

ricardoV94 committed:
Remove "distribution" from test class names

ricardoV94 commented:
I pushed a change to your PR to make the tests names more succinct and readable. Don't forget to pull before committing new changes

matteo-pallini committed:
Add discrete Weibull, improve Beta and some minor refactoring

matteo-pallini committed:
Fix typos in checks naming and add sanity check

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -485,43 +399,391 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+            self.pymc_rv.eval(), self.reference_dist_draws, decimal=self.decimal
+        )
 
+    def check_pymc_params_match_rv_op(self) -> None:
```
Why the return `-> None` only in this one?

matteo-pallini replied:
Removed. Initially I added typing more thoroughly, which then got lost with refactoring

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -485,43 +399,391 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+        p = seeded_numpy_distribution_builder("uniform")
+        return (
+            lambda size, q, beta: np.ceil(
+                np.power(np.log(1 - p(self)(size=size)) / np.log(q), 1.0 / beta)
```
I wonder whether this can be simplified / made more readable? the `p(self)(size=size)` looks very opaque

matteo-pallini replied:
yeah, it was not great. I did a minor change, it should be better now. Let me know if you have a better alternative in mind

ricardoV94 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -485,43 +399,391 @@ class TestMoyal(BaseTestCases.BaseTestCase):
+    repeated_params_shape = 5
+
+    def test_distribution(self):
+        self.sanity_check_tests_to_run()
```
Maybe rename to something like `self.validate_tests_list`?

ricardoV94 reviewed (commented):
Left some minor comments. The way the `discrete_weibul_rng_fn` is implemented looks a bit complicated. Do you think it might read better if we unpack that lambda to an explicit function?

Otherwise things LGTM!

ricardoV94 reviewed (approved):
LGTM, much more readable and obvious how to apply to new distributions going into the future!

ricardoV94 commented:
Thanks a lot @drabbit17. Looking forward to your next PR!

matteo-pallini commented:
Thanks for reviewing and helping in making the final version much better
